### PR TITLE
Adding new API HOWTO

### PR DESCRIPTION
**Pull request by beldmit**
A very brief explanation of how to add custom functions to OpenSSL.
Inspired by Rich Salz's letter to openssl-users@

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: HACKING [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ If you are adding new public functions to the custom library build, you need to
+ either add a prototype in one of the existing OpenSSL header files;
+ or provide a new header file and edit Configurations/unix-Makefile.tmpl to pick up that file.
+ 
+ After that performthe following steps:
+ 
+ ./configure [your-options]
+ make update
+ make
+
```
Comment (tmshort) [inline HACKING]: Add `make doc-nits` to make sure documentation is correct.

Comment (beldmit) [reply HACKING]: before or after make?

Comment (tmshort) [reply HACKING]: I personally do it after the make; but I'm not sure it matters. It should be done after `make update`, however.

File: HACKING [hunk 2]
```diff
@@ -0,0 +1,20 @@
+ MODIFYING OPENSSL SOURCE
+ ------------------------
+ This document describes the way to add custom modifications to OpenSSL sources.
+ 
+ If you are adding new public functions to the custom library build, you need to
+ either add a prototype in one of the existing OpenSSL header files;
+ or provide a new header file and edit Configurations/unix-Makefile.tmpl to pick up that file.
+ 
+ After that performthe following steps:
```
Comment (slontis) [inline HACKING]: space missing between perform the

Comment (paulidale) [inline HACKING]: missing space between _perform_ and _the_.

File: HACKING [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ 
+ ./configure [your-options]
+ make update
+ make
+ 
+ “make update” ensures that your functions declarations are added to util/libcrypto.num or util/libssl.num
+ If you plan to submit the changes you made to OpenSSL (see CONTRIBUTING), it's worth running
```
Comment (paulidale) [inline HACKING]: Colon at end of line so it's consistent with the above.

File: HACKING [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ ------------------------
+ This document describes the way to add custom modifications to OpenSSL sources.
+ 
+ If you are adding new public functions to the custom library build, you need to
+ either add a prototype in one of the existing OpenSSL header files;
+ or provide a new header file and edit Configurations/unix-Makefile.tmpl to pick up that file.
+ 
+ After that performthe following steps:
+ 
+ ./configure [your-options]
```
Comment (slontis) [inline HACKING]: The options '-Werror --strict-warnings' are probably a good thing to add..

Comment (paulidale) [inline HACKING]: Indent these lines perhaps?

File: HACKING [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ make update
+ make
+ 
+ “make update” ensures that your functions declarations are added to util/libcrypto.num or util/libssl.num
+ If you plan to submit the changes you made to OpenSSL (see CONTRIBUTING), it's worth running 
+
+ make doc-nits
```
Comment (paulidale) [inline HACKING]: Indent this one too.

File: HACKING (hunk 3 above)
Comment (slontis) [inline HACKING]: Maybe it also needs to mention:-
make update also generates files related to errors.
If a merge error occurs in one of these generated files then the generated files need to be removed
and regenerated using make update. To aid in this process the generated files can be committed seperately so they can be removed easily.

**Comment by beldmit** [comment]
Oops. It was not in my repo?..

**Pull request closed without merging**